File: luogu/p1378.cpp
#include<bits/stdc++.h>
using namespace std;

int n;
double xx;
double yy;
double xxx;
double yyy;
const double PI = 3.1415926535;
double x[10], y[10];
double ans, maxv, r[10];
bool used[10];

double calc(int i)
{
    double d1 = min(abs(x[i]-xxx), abs(x[i]-xx));
    double d2 = min(abs(y[i]-yyy), abs(y[i]-yy));
    double d = min(d1, d2);
    for(int j = 0; j < n; j++)
    {
        if(j == i || !used[j]) continue;
        double temp = sqrt((x[i]-x[j])*(x[i]-x[j])+(y[i]-y[j])*(y[i]-y[j]));
        d = min(d, max(temp-r[j], 0.0));
    }
    return d;
}

void dfs(int k, double sums)
{
    if(k > n)
    {
        maxv = max(sums, maxv);
        return;
    }
    for(int i = 0; i < n; i++)
    {
        if(!used[i])
        {
            r[i] = calc(i);
            used[i] = true;
            dfs(k+1, sums+PI*r[i]*r[i]);
            used[i] = false;
        }
    }
}

int main()
{
    cin >> n;
    cin >> xx >> yy >> xxx >> yyy;
    for(int i = 0; i < n; i++)
        cin >> x[i] >> y[i];
    maxv = 0;
    dfs(1, 0);
    cout << int(abs(xxx-xx)*abs(yyy-yy)-maxv+0.5) << endl;
    return 0;
}

File: luogu/p1378.py
import math

n = int(input())

x1, y1, x2, y2 = map(float, input().split())

point = []
used = [False]*n
r = [0.0]*n

for i in range(n):
    x, y = map(float, input().split())
    point.append((x, y))

max_v = 0.0


def calDistance(i, j):
    return math.sqrt((point[i][0]-point[j][0]) ** 2 + (point[i][1]-point[j][1]) ** 2)


def findMax(i):
    d1 = min(abs(point[i][0]-x1), abs(point[i][0]-x2))
    d2 = min(abs(point[i][1]-y1), abs(point[i][1]-y2))
    d = min(d1, d2)
    for j in range(n):
        if(j != i and used[j] == True):
            d = min(d, max(calDistance(i, j)-r[j], 0.0))
    return d


def dfs(p, sums):
    global max_v
    if(p >= n):
        max_v = max(max_v, sums)
        return
    for i in range(n):
        if(used[i] == True):
            continue
        used[i] = True
        r[i] = findMax(i)
        dfs(p+1, sums+r[i]*r[i]*math.pi)
        used[i] = False


dfs(0, 0.0)

ans = abs(x2-x1)*abs(y1-y2)-max_v
print(int(ans+0.5))
